drivers/modem/gsm_ppp.c

```diff
@@ -944,42 +968,48 @@ static void gsm_configure(struct k_work *work)
 		return;
 	}
 
-	if (IS_ENABLED(CONFIG_GSM_MUX) && ret == 0 &&
-	    gsm->mux_enabled == false) {
-		gsm->mux_setup_done = false;
-
-		ret = mux_enable(gsm);
-		if (ret == 0) {
-			gsm->mux_enabled = true;
-		} else {
-			gsm->mux_enabled = false;
-			(void)k_work_reschedule(&gsm->gsm_configure_work,
-						K_NO_WAIT);
-			return;
-		}
-
-		LOG_DBG("GSM muxing %s", gsm->mux_enabled ? "enabled" :
-							    "disabled");
+	gsm->state = GSM_PPP_AT_RDY;
 
-		if (gsm->mux_enabled) {
-			gsm->state = STATE_INIT;
+#if IS_ENABLED(CONFIG_GSM_MUX)
+	if (mux_enable(gsm) == 0) {
+		gsm->state = GSM_PPP_MUX_ENABLED;
+	} else {
+		(void)k_work_reschedule(&gsm->gsm_configure_work,
+					K_NO_WAIT);
+		return;
+	}
 
-			k_work_init_delayable(&gsm->gsm_configure_work,
-					      mux_setup);
+	LOG_DBG("GSM muxing enabled");
+	gsm->state = GSM_PPP_STATE_INIT;
 
-			(void)k_work_reschedule(&gsm->gsm_configure_work,
-						K_NO_WAIT);
-			return;
-		}
-	}
+	k_work_init_delayable(&gsm->gsm_configure_work,
+			      mux_setup);
 
+	(void)k_work_reschedule(&gsm->gsm_configure_work,
+				K_NO_WAIT);
+#else
+	gsm->state = GSM_PPP_SETUP;
 	gsm_finalize_connection(gsm);
+#endif
 }
 
 void gsm_ppp_start(const struct device *dev)
 {
 	struct gsm_modem *gsm = dev->data;
 
+	if (gsm->state != GSM_PPP_STOP) {
+		LOG_ERR("gsm_ppp is already started");
```

Comment:
Well now that you have a machinery for tracking state completely, changing the prototype of start/stop to return `int` would make sense imo. There's currently a problem where stopping gsm_ppp is hazarduous if the configuration loop is currently running - making start/stop return error codes would be a huge mitigation.